### ISSUE TITLE
Runtime Error: No key found in the acoustic query map...

### ISSUE THREAD

**Issue by hulkmeistercoder** [microsoft/ProjectAcoustics#95]
I'm receiving this error and having some trouble deciphering it:

"LogAcousticsRuntime: Error: No key found in the acoustic query map for source:-536870912. This most likely means this source did not register first (RegisterSourceObject) before updating."

I do not know what the latter means as I followed the integration instructions and there was no mention of a RegisterSourceObject step.

Can you give some insight on this?

Thanks!

**Comment by KyleStorck** [comment]
Hi @hulkmeistercoder. Are you using the marketplace plugin for Unreal Audio, or are you using the Unreal/Wwise offering from the download center?

**Comment by hulkmeistercoder** [comment]
@KyleStorck I am using the marketplace plugin for Unreal Audio.

**Comment by KyleStorck** [comment]
Interesting. That error message means that the plugin has entered a weird state. Doesn't necessarily mean you missed a step. Some questions to give me more info...

- You went through all the steps here? https://docs.microsoft.com/en-us/gaming/acoustics/unreal-audio-integration#audio-setup-in-unreal. (Set up acoustics space actor, selected Project Acoustics as Source Data Override plugin, added an Attenuation Settings object to the sound source and enabled Source Data Override)
- Do you see that error message once or does it get spammed. Any other error messages in the OutputLog? Can you post pics of the Output Log.
- Can you tell me more about your sound source. What sound asset are you using? How does it get played (play on start, or does it get triggered). Can you post a pic of the Details pane of your sound source.

**Comment by hulkmeistercoder** [comment]
@KyleStorck I'm working on getting you the details you asked for above, but re. question 2, I'm sometimes receive this other message:

"LogAcousticsRuntime: Warning: No acoustic query result found for source:4. This most likely means a background query did not complete in time."

**Comment by hulkmeistercoder** [comment]
@KyleStorck here are the details:
1a. Created the Acoustics Space actor.
![ProjectAcoustics_AcousticsSpaceActor](https://user-images.githubusercontent.com/65577468/170845492-cde83d16-8574-4980-9ba8-eab9a4a12518.png)
)
1b. Setup the override plugin.
![ProjectAcoustics_ProjectSettings_WindowsAudio](https://user-images.githubusercontent.com/65577468/170845514-c5e7bfd4-e88f-4fdd-9e27-83fa3d0af05d.png)
1c. Created an attenuation setting.
![ProjectAcoustics_AttenuationSettings](https://user-images.githubusercontent.com/65577468/170845601-651fddd9-697d-4fd8-a6e7-d6e9c07d6536.png)
1d. Created an audio component with replication and attenuation settings (this is different from my intended setup, but still produces the same issue).
![ProjectAcoustics_AudioComponentSettings](https://user-images.githubusercontent.com/65577468/170845529-34a9a720-261a-433b-a84e-0d8cb03c0256.png)
2. The message gets spammed over and over. With my simplified audio component setup, I have the deactivation of the component bound to a key press. The errors only start after I stop the audio component, and stop when I reactivate it.

The warnings on the other hand occur sporadically whether the component is on or off. Here's a picture of the log like you requested:
![ProjectAcoustics_OutputLog](https://user-images.githubusercontent.com/65577468/170845651-26651c9e-f0e8-4ae4-a8ce-59e3891e4009.png)
3. To remove the possibility of a third-party plugin causing the issue, I removed the plugin and opted to play the sound source as a looping audio component as per the instructions. The picture is above with the first question. My audio source is just a Sound Wave with the attenuation settings configured on it. Details panel as requested:

![ProjectAcoustics_SoundWave_TopHalf](https://user-images.githubusercontent.com/65577468/170845787-f4f6f540-1579-4ca8-8bd1-e7c3f6a7d2c1.png)
![ProjectAcoustics_SoundWave_BottomHalf](https://user-images.githubusercontent.com/65577468/170845788-2b81702f-343b-42b6-b79d-763350bb5390.png)

**Comment by hulkmeistercoder** [comment]
Since the game is multiplayer, and my acoustics data did have a navmesh dependency, I went ahead and enabled "Allow Client Side Navigation". It did not fix the issue.

**Comment by hulkmeistercoder** [comment]
The only place I'm having trouble peering into is the ACE file. Is it possible the error has something to do with this file? Is there a way to visualize this data to see if that's the case?

**Comment by KyleStorck** [comment]
Thanks for the great info! Your setup looks fine. No, this wouldn't be a problem with the ACE file. We implemented a new threading system that runs the acoustic queries on a background thread. That is where the warnings/errors are coming from.

 Couple more questions.

Roughly speaking, what are the specs of your PC? Is it a powerful machine or less powerful? Is there a lot going on in your scene or not really? Those warning messages sometimes pop up when there is a lot going on, but they shouldn't pop up this much on a normal machine.

That error message being spammed after the sound is stopped is unusual. Do they keep going forever, or do they stop after a while? How are you "deactivating the component"?

**Comment by hulkmeistercoder** [comment]
@KyleStorck here's a snippet from my dxdiag:

```
------------------
System Information
------------------
      Time of this report: 5/28/2022, 17:06:08
             Machine name: 
               Machine Id: 
         Operating System: Windows 10 Home 64-bit (10.0, Build 19044) (19041.vb_release.191206-1406)
                 Language: English (Regional Setting: English)
      System Manufacturer: System manufacturer
             System Model: System Product Name
                     BIOS: BIOS Date: 10/21/20 15:37:40 Ver: 05.00011 (type: BIOS)
                Processor: AMD Ryzen 9 3950X 16-Core Processor             (32 CPUs), ~3.7GHz
                   Memory: 65536MB RAM
      Available OS Memory: 65442MB RAM
                Page File: 25620MB used, 49549MB available
              Windows Dir: C:\Windows
          DirectX Version: DirectX 12
      DX Setup Parameters: Not found
         User DPI Setting: 96 DPI (100 percent)
       System DPI Setting: 96 DPI (100 percent)
          DWM DPI Scaling: Disabled
                 Miracast: Available, with HDCP
Microsoft Graphics Hybrid: Not Supported
 DirectX Database Version: 1.0.8
           DxDiag Version: 10.00.19041.0928 64bit Unicode
```

```
---------------
Display Devices
---------------
           Card name: AMD Radeon RX 5600 XT
        Manufacturer: Advanced Micro Devices, Inc.
           Chip type: AMD Radeon Graphics Processor (0x731F)
            DAC type: Internal DAC(400MHz)
         Device Type: Full Device (POST)
          Device Key: Enum\PCI\VEN_1002&DEV_731F&SUBSYS_51241849&REV_CA
       Device Status: 0180200A [DN_DRIVER_LOADED|DN_STARTED|DN_DISABLEABLE|DN_NT_ENUMERATOR|DN_NT_DRIVER] 
 Device Problem Code: No Problem
 Driver Problem Code: Unknown
      Display Memory: 38824 MB
    Dedicated Memory: 6103 MB
       Shared Memory: 32720 MB
        Current Mode: 2560 x 1440 (32 bit) (60Hz)
         HDR Support: Not Supported
    Display Topology: Extend
 Display Color Space: DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709
     Color Primaries: Red(0.661133,0.332031), Green(0.301758,0.613281), Blue(0.149414,0.059570), White Point(0.313477,0.329102)
   Display Luminance: Min Luminance = 0.500000, Max Luminance = 270.000000, MaxFullFrameLuminance = 270.000000
        Monitor Name: Generic PnP Monitor
       Monitor Model: DELL U2715H
          Monitor Id: DELD065

```

**Comment by hulkmeistercoder** [comment]
@KyleStorck the messages go on seemingly forever. Here's a snippet of how I stop the audio on keypress. This is under a Character actor. The audio component starts out auto-activated.
![ProjectAcoustics_AudioStopMethod](https://user-images.githubusercontent.com/65577468/170846740-ecaa050b-1010-4399-835f-3a3fb2efe9b3.PNG)

**Comment by hulkmeistercoder** [comment]
@KyleStorck re. whether there's a lot going on in a scene, here's an example of the arena we're in. The red lines are where we expect players to go, and where we want the acoustics to be active:
![ProjectAcoustics_SceneClutter](https://user-images.githubusercontent.com/65577468/170846887-2ff669bc-3680-4855-9304-da1cbd36b8e7.png)

**Comment by KyleStorck** [comment]
Ok, thanks for the info. I will start by trying out stopping audio the way you do. It's possible that's different than I how I've tested and I might not be handling it correctly. I'm still stumped about that warning though. Surprised your seeing that on your machine. I'll keep thinking about that. I'll start on this when I get back to the office on Tuesday.

**Comment by hulkmeistercoder** [comment]
@KyleStorck thank you for being so responsive! Let me know how I should be stopping the audio. I am not opposed to handling it another way. I'm just trying to get a proof of concept without the errors running.

Thanks again!

**Comment by hulkmeistercoder** [comment]
I don't remember how I enabled this, but this debug info showed up on my screen that seems to be associated with the error:
![ProjectAcoustics_MoreDebugInfo](https://user-images.githubusercontent.com/65577468/170847834-86328c22-e41e-46c1-a0b3-b0007bb1d011.png)

**Comment by KyleStorck** [comment]
Those acoustic parameter errors shouldn't be related to the other ones. It's normal for those errors to show up occasionally. They sometimes show up if the sound source goes inside a voxel. Do those errors show up all the time? If those queries always fail, that's indicative of something wrong with your ACE.

**Comment by hulkmeistercoder** [comment]
The errors don't show up all the time. It happens intermittently depending on how I'm facing the other player. Whether they're perfectly aligned with the query errors is hard to determine.

I did perform a local bake with the setup I had. Since you mentioned something may be wrong with the ACE file, I performed a re-bake using Azure batch. Those errors are still showing up with the new file.

So, I look forward to what you find after Tuesday with more testing.

**Comment by KyleStorck** [comment]
The consequences for the on-screen probe interp errors and the "query did not complete in time" warning are that new acoustic parameters are not retrieved for that frame. In those cases, the last known good parameters will still be used. All the acoustic features are still working, they just may be delayed a frame or two.

**Comment by hulkmeistercoder** [comment]
@KyleStorck setup a test with the following:
1. Stripped down version of our game character.
2. One audio component turned on/off in the same way mentioned previously.
3. Level ambience off.
4. Playing in PIE mode with 2 clients.

Also, the query errors do not happen in PIE mode with only 1 client nor in standalone.

The videos have chapters for you to jump around to the important events for when warnings/errors start. Gameplay UI has the UTC timestamp and the state of the audio component at various points. Let me know if you need the logs just pasted here.

https://youtu.be/BE63tnGbRls Associated with Debug Log v1:
[ProjectAcoustics_GithubIssue95_DebugLog_v1.txt](https://github.com/microsoft/ProjectAcoustics/files/8819147/ProjectAcoustics_GithubIssue95_DebugLog_v1.txt)

https://youtu.be/gS3nvXJsYU0 Associated with Debug Log v2: [ProjectAcoustics_GithubIssue95_DebugLog_v2.txt](https://github.com/microsoft/ProjectAcoustics/files/8819152/ProjectAcoustics_GithubIssue95_DebugLog_v2.txt)

**Comment by KyleStorck** [comment]
Thanks for the great report @hulkmeistercoder. I'm still parsing through it, but like you said, those errors do appear to have to do with running multiple clients. We had not tested that scenario yet. I am less worried about the warnings now. Those appear to be sporadic enough as to not cause any major problems. I will focus on the errors now.

I need to get this error reproducing on my system. Do you know of any multiplayer sample content that can help me quickly get something similar to your multiplayer game set up? If not, I'll figure something out. Maybe I can get multiple clients running in the Lyra demo scene.

**Comment by hulkmeistercoder** [comment]
@KyleStorck I think Lyra is a good place to start. Our multiplayer game mode is based on AGameMode, which does things similar to Lyra's. I'll keep looking at the differences between their mode and ours to see if there's anything that could affect reproducing this issue; and call it out, if I see anything.

The plugin seems to be working quite nicely in multi-client mode despite the errors.

**Comment by KyleStorck** [comment]
Hi @hulkmeistercoder. We looked at this closer and realized that running in editor with multiple clients is not a scenario that we support and is not something that is easily fixed. Running with multiple clients in one editor is essentially the same as having multiple listeners, and we don't have an established solution for that yet. 

But after reproing myself, I can confirm that the errors and warnings you saw in OutputLog are due to the multi-client scenario. You should not see those errors/warnings when running in standalone as a single client.

The probe-interp errors you see in the debug view are a separate issue. Those errors can be normal. Normally they happen if the sound source is within a voxel. If you see those errors a lot, you can open a separate issue and we can help you debug those.

**Comment by hulkmeistercoder** [comment]
@KyleStorck thanks for taking a closer look and helping me understand it better!

There are a couple questions I'd do some testing for, but I'm blocked by some unrelated cooking/packaging issues. So, to know what to expect when those are resolved, I have the following questions:
1. If I were to run each client in their own process on the same machine, I should expect to not see these errors?
2. Re. packaging for a dedicated server, is Project Acoustics a runtime requirement server-side?

**Comment by KyleStorck** [comment]
Yes, if each client has their own process, that should work fine.

I don't know much about how clients/servers work in Unreal, but if there is no audio being played on the server, then Project Acoustics is not needed there.

**Comment by hulkmeistercoder** [comment]
Okay, thanks again!
